### PR TITLE
Review of HOPE-6

### PR DESCRIPTION
**Pull request by jay-tyler**
Proposed API refactor and a question about precedence of defaults.

### CODE REVIEW

File: all/HOPE-6--Per-Interface-Defaults.md [hunk 1]
```diff
@@ -18,14 +18,16 @@ This HOPE details the addition of per-interface default parameter support to Hug
...
+Each interface decorator would allow addition of arbitrary kwargs to pass as defaults to the api function, should
+the api function return None.
+
+JMT TODO: Is the description above accurate? Or maybe None is allowed, but interface defaults always take precedence over functional defaults.
 
 Looking somewhat like this:
 
 
 ```python
-@hug.http(defaults={'user': 'system'})
```
Comment (jay-tyler) [inline all/HOPE-6--Per-Interface-Defaults.md:28]: Motivation was to remove mutable data types as default kwarg vals.

File: all/HOPE-6--Per-Interface-Defaults.md [hunk 2]
```diff
@@ -36,7 +38,7 @@ The defaults would also be able to be applied to parameters whose endpoints have
 
 ```python
 @hug.http()
-@hug.cli(defaults={'user': getpass.getuser()})
+@hug.cli(user=getpass.getuser())
```
Comment (jay-tyler) [inline all/HOPE-6--Per-Interface-Defaults.md:41]: Another option would be to use a decorator namespace, e.g. `@hug.cli.default(user=getpass.getuser())`.

Not yet sure which option is better. `hug.cli.default` might be more testable, and might be easier to document (i.e. use any kwargs you want). That said, it's implying addition of an interface. Might need to accommodate use of both `@hug.cli` and `@hug.cli.default`

File: all/HOPE-6--Per-Interface-Defaults.md [hunk 3]
```diff
@@ -18,14 +18,16 @@ This HOPE details the addition of per-interface default parameter support to Hug
 
 ## Implementation
 
-Each interface decorator would add a new, optional, defaults argument. This argument would take a dictionary of
-parameter names and their intended defaults per the interface.
+Each interface decorator would allow addition of arbitrary kwargs to pass as defaults to the api function, should
+the api function return None.
+
+JMT TODO: Is the description above accurate? Or maybe None is allowed, but interface defaults always take precedence over functional defaults.
```
Comment (jay-tyler) [inline all/HOPE-6--Per-Interface-Defaults.md:24]: Question for @timothycrosley

Comment (timothycrosley) [reply all/HOPE-6--Per-Interface-Defaults.md:24]: If I understand the question correctly: I think that it's the latter, an api function can have it's own defaults but the interface defaults will take precedence if called from the specific interface.

**Pull request merged**